ores/scoring_systems/celery_queue.py

```diff
@@ -0,0 +1,287 @@
+                                      if injection_caches is not None else None
+                    task_id = context.format_id_string(
+                        model_name, rev_id, injection_cache=injection_cache)
+                    self.application.backend.store_result(
+                        task_id, {}, 'REQUESTED')
+
+    def _score(self, *args, **kwargs):
+        self._check_queue_full()
+        return super()._score(*args, **kwargs)
+
+    def _check_queue_full(self):
+        # Check redis to see if the queue of waiting tasks is too big.
+        # This is a hack to implement backpressure because celery doesn't
+        # support it natively.
+        # This will result in a race condition, but it should have OK
+        # properties.
+        if self.redis is not None:
+            queue_size = self.redis.llen(DEFAULT_CELERY_QUEUE)
+            if queue_size > self.queue_maxsize:
+                message = "Queue size is too full {0}".format(queue_size)
+                logger.warning(message)
+                raise ScoreProcessorOverloaded(message)
+
+    @classmethod
+    def _build_context_map(cls, config, name, section_key="scoring_systems"):
+        from .. import ores
+        from ..scoring_context import ScoringContext, ClientScoringContext
+
+        section = config[section_key][name]
+
+        if hasattr(ores, "_is_wsgi_client") and ores._is_wsgi_client:
+            ScoringContextClass = ClientScoringContext
+        else:
+            ScoringContextClass = ScoringContext
+
+        return {name: ScoringContextClass.from_config(config, name)
+                for name in section['scoring_contexts']}
+
+    @classmethod
+    def from_config(cls, config, name, section_key="scoring_systems"):
+        logger.info("Loading CeleryQueue '{0}' from config.".format(name))
+        section = config[section_key][name]
+
+        kwargs = cls._kwargs_from_config(
+            config, name, section_key=section_key)
+        queue_maxsize = section.get('queue_maxsize')
+        application = celery.Celery(__name__)
+        application.conf.update(**{k: v for k, v in section.items()
+                                   if k not in ('class', 'context_map',
+                                                'score_cache',
+                                                'metrics_collector', 'timeout',
+                                                'queue_maxsize')})
+        application.conf.CELERY_CREATE_MISSING_QUEUES = True
+
+        return cls(application=application,
+                   queue_maxsize=queue_maxsize, **kwargs)
+
+
+PASS_HOST_PORT = re.compile(
+    r"((?P<password>[^@]+)@)?" +
+    r"(?P<host>[^:]+)?" +
+    r"(:(?P<port>[0-9]+))?"
+)
+"""
+Matches <password>@<host>:<port>
+"""
+
+
+def redis_from_url(url):
+    """
+    Converts a redis URL used by celery into a `redis.Redis` object.
+    """
+    # Makes sure that we only try to import redis when we need
+    # to use it
```

Comment:
Is it ever not used? This seems like a good candidate to split out, since it's a different level of abstraction that could benefit from polymorphism